### PR TITLE
[MOE] redirect moe_env from global_variables to core

### PR DESCRIPTION
**Pull request by 1SAA**
- use dist process group instead of ParallelMode in moe layer.py
- add comments for experts construction functions

### CODE REVIEW

File: colossalai/nn/layer/moe/_operation.py [hunk 1]
```diff
@@ -4,11 +4,11 @@
 from typing import Any, Tuple, Optional
 from torch.distributed import ProcessGroup
 
-U_CUDA_MODE = False
+KERNEL_OPTIM = False
```
Comment (feifeibear) [inline colossalai/nn/layer/moe/_operation.py:7]: USE_KERNEL_OPTIM

File: colossalai/nn/layer/moe/experts.py [hunk 2]
```diff
@@ -2,18 +2,24 @@
...
 
 
 class MoeExperts(nn.Module):
+    """Basic class for experts in MoE. It stores what kind of communication expersts use
+    to exchange tokens, how many experts in a single GPU and parallel information such as
+    expert parallel size, data parallel size and their distributed communication groups.
+    """
 
-    def __init__(self, comm: str):
+    def __init__(self, comm: str, num_experts: int):
```
Comment (feifeibear) [inline colossalai/nn/layer/moe/experts.py:16]: comm -> comm_name

File: colossalai/nn/layer/moe/_operation.py (hunk 1 above)
Comment (1SAA) [reply colossalai/nn/layer/moe/_operation.py:7]: Whether uses kernel is controled by use_kernel in MoeLayer. KERNEL_OPTIM only shows whether CUDA kernel optimization can be used.

File: colossalai/nn/layer/moe/experts.py [hunk 3]
```diff
@@ -29,53 +35,48 @@ class Experts(MoeExperts):
...
+        with seed(ParallelMode.TENSOR):
+            self.experts = nn.ModuleList([expert(**expert_args) for _ in range(self.num_local_experts)])
 
+        # Attach parallel information for all parameters in Experts
         for exp in self.experts:
             for param in exp.parameters():
-                param.__setattr__('moe_param', True)
-
-        self.num_local_experts = num_local_experts
+                param.__setattr__('moe_info', self.dist_info)
```
Comment (feifeibear) [inline colossalai/nn/layer/moe/experts.py:47]: Eevery param stores the duplicated dist_info?

File: colossalai/nn/layer/moe/experts.py [hunk 4]
```diff
@@ -29,53 +35,48 @@ class Experts(MoeExperts):
...
 
     def forward(self, inputs):
+        # Split inputs for each expert
```
Comment (feifeibear) [inline colossalai/nn/layer/moe/experts.py:50]: inputs should a single Tensor?
Did chunk works for a list of Tensors?

If so
assert isinstance(inputs, torch.Tensor)

File: colossalai/nn/loss/loss_moe.py [hunk 5]
```diff
@@ -1,7 +1,7 @@
 import torch.nn as nn
 from colossalai.registry import LOSSES
 from torch.nn.modules.loss import _Loss
-from colossalai.global_variables import moe_env
+from colossalai.core import moe_context as moe_env
```
Comment (feifeibear) [inline colossalai/nn/loss/loss_moe.py:4]: I let you change the name of global variable in previous PR
Use upper case and GLOBAL_ as a prefix.

File: colossalai/nn/layer/moe/_operation.py (hunk 1 above)
Comment (feifeibear) [reply colossalai/nn/layer/moe/_operation.py:7]: you name is misleading. I suppose KERNEL_OPTIM is an optimizer instance. Change the name and make it explains itself.

File: colossalai/nn/layer/moe/experts.py (hunk 3 above)
Comment (1SAA) [reply colossalai/nn/layer/moe/experts.py:47]: Python object only stores a reference. The dist_info is returned from moe_context information dictionary, there is no duplicated dist_info.

File: colossalai/nn/layer/moe/_operation.py (hunk 1 above)
Comment (1SAA) [reply colossalai/nn/layer/moe/_operation.py:7]: I use COL_MOE_KERNEL instead.

File: colossalai/nn/layer/moe/experts.py (hunk 2 above)
Comment (1SAA) [reply colossalai/nn/layer/moe/experts.py:16]: done

File: colossalai/nn/layer/moe/experts.py (hunk 4 above)
Comment (1SAA) [reply colossalai/nn/layer/moe/experts.py:50]: added hint, torch.Tensor, for inputs

File: colossalai/nn/loss/loss_moe.py (hunk 5 above)
Comment (1SAA) [reply colossalai/nn/loss/loss_moe.py:4]: used MOE_CONTEXT instead

File: colossalai/nn/layer/moe/_operation.py (hunk 1 above)
Comment (feifeibear) [reply colossalai/nn/layer/moe/_operation.py]: COL_MOE_KERNEL_FLAG? it's a bool var. you name should reflect its type.

File: colossalai/nn/layer/moe/experts.py (hunk 3 above)
Comment (feifeibear) [reply colossalai/nn/layer/moe/experts.py:47]: I see

File: colossalai/nn/layer/moe/_operation.py (hunk 1 above)
Comment (1SAA) [reply colossalai/nn/layer/moe/_operation.py]: done

**Review by feifeibear — APPROVED** [review]
